Automattic/jetpack
# Boost: Add EmptyCSSError Critical CSS error support

This is a port of a PR from the legacy Jetpack Boost repository.

#### Changes proposed in this Pull Request:

~~It works with the [Jetpack Boost Critical CSS Gen library "Throw an error if Critical CSS is empty"](https://github.com/Automattic/jetpack-boost-critical-css-gen/pull/19) PR.~~

This PR introduce an new Critical CSS error (`EmptyCSSError`) descriptor which handle the case where the Critical CSS could not be generated because no external CSS was found in a specific URL.

~~**IMPORTANT**: If we decide to merge these 2 PRs then we need to merge first  [Jetpack Boost Critical CSS Gen library PR](https://github.com/Automattic/jetpack-boost-critical-css-gen/pull/19) PR and then revert the changes made to the `package.json` and `pnpm-lock.yaml` files.~~


#### Jetpack product discussion


N/A

#### Does this pull request change what data or activity we track or use?


N/A

#### Testing instructions:






This is actually difficult to test somehow. The issue became apparent while testing Jetpack Boost plugin on a [LocalWP](https://localwp.com/) setup using their live link feature (ngrok tunneling to the LocalWP site). To replicate on another local environment it would require to use a Basic Authentication plugin and a theme using only external CSS. 

Alternatively to "fake" it on a local environment the following can be done:
- Replace lines https://github.com/Automattic/jetpack-boost/blob/feature/add-empty-css-error-descriptor/app/modules/critical-css/class-critical-css.php#L749-L751 by

```
if ( empty( $parsed['host'] ) ) {
	return $src;
}
```

- Replace line https://github.com/Automattic/jetpack-boost/blob/feature/add-empty-css-error-descriptor/app/modules/critical-css/class-critical-css.php#L793
by 
```
echo '';
```

**NOTE**: In a way it is an edge case, but I think it is probably a good idea to catch error with URLs for which Critical CSS could not be created because no CSS were found.

See Video of the issue on a LocalWP site (using live link): 

https://user-images.githubusercontent.com/927932/125466935-fd9514db-d77b-4409-9175-c9aac0de3bcd.mov

You'll notice that it says the Critical CSS was generated but in fact there is no Critical CSS.

Below are screenshots showing how it will looks like with the fix on a local environment (modified to recreate the LocalWP environment):

<img width="845" alt="Screen Shot 2021-07-30 at 4 42 44 PM" src="https://user-images.githubusercontent.com/927932/127619146-09947d7f-3787-46f0-b6fa-b2a0ff9a6ceb.png">
<img width="889" alt="Screen Shot 2021-07-30 at 4 43 00 PM" src="https://user-images.githubusercontent.com/927932/127619151-6de8db24-4ee9-4dd2-bf0a-4d30fd3c4838.png">

**NOTE**: TODO: We might want to update the suggestion description for the error.
**NOTE**: TODO: Before merging it still needs to be tested with a LocalWP environment.
